chore: included subrelation witness degrees in the relations relevant to zk-sumcheck

iakovenkos opened the pull request:
The subrelation witness degrees are used in zk-sumcheck and affect the sizes of the univariates in the SumcheckTupleOfTuplesOfUnivariates as well as the sizes of the sumcheck round univariates.

In several relations, I used upper bounds on the witness degrees which are simply given by the corresponding subrelation's length - 1. Indicated explicitly that those are upper bounds. Computed the precise constants for the relations relevant to Mega. 

The auxiliary and logderiv_lookup_relation contain a trick allowing to re-use the values accumulated by an accumulator of a different size using ShortView.

iakovenkos committed:
chore: added subrelation witness degrees to relations relevant to zk-sumcheck

maramihali commented on barretenberg/cpp/src/barretenberg/relations/ultra_arithmetic_relation.hpp:
```diff
@@ -12,6 +12,12 @@ template <typename FF_> class UltraArithmeticRelationImpl {
     };
 
+    /**
+     * @brief Total degrees of sub-relations considered as polynomials in witnesses.
```
```suggestion
     * @brief Total degrees of subrelations considered as polynomials in witnesses.
```

maramihali replied:
Could you rephrase this comment so it's clear where it's being used? Also I think it's quite a confusing comment. Here and in the other places

maramihali commented on barretenberg/cpp/src/barretenberg/relations/translator_vm/translator_permutation_relation.hpp:
```diff
@@ -13,7 +13,14 @@ template <typename FF_> class TranslatorPermutationRelationImpl {
+     *
+     */
+    static constexpr std::array<size_t, 2> SUBRELATION_WITNESS_DEGREES{
+        6, // grand product construction sub-relation
```
no need for `-` in subrelation :)

maramihali commented on barretenberg/cpp/src/barretenberg/relations/databus_lookup_relation.hpp:
```diff
@@ -57,6 +57,17 @@ template <typename FF_> class DatabusLookupRelationImpl {
+     * @brief Upper bound on total degrees of sub-relations considered as polynomials in witnesses.
+     *
+     */
+    static constexpr std::array<size_t, NUM_BUS_COLUMNS * 2> SUBRELATION_WITNESS_DEGREES{
```
please explain here the -1 and in the other places, it really helps the reader in a few months

maramihali commented on barretenberg/cpp/src/barretenberg/relations/auxiliary_relation.hpp:
```diff
@@ -97,8 +111,9 @@ template <typename FF_> class AuxiliaryRelationImpl {
-        using Accumulator = typename std::tuple_element_t<0, ContainerOverSubrelations>;
+        using Accumulator = typename std::tuple_element_t<3, ContainerOverSubrelations>;
         using View = typename Accumulator::View;
+        using ShortView = typename std::tuple_element_t<0, ContainerOverSubrelations>::View;
```
I think it would be nice to add the related comment from the PR description here

maramihali reviewed (approved):
Overall looks good, thanks for the additional docs, just some nit picks and requests for clarifying comments.

iakovenkos committed:
resolved comments

iakovenkos committed:
Merge branch 'master' into zk-relation-length-adjustments

iakovenkos commented:
Resolved all isssues: added clearer definitions for SUBRELATION_WITNESS_DEGREES and explained the upper bounds appearing in some of the relations (mostly ECCVM and Translator). Added comments for the ShortView trick in logderivative and auxiliary relations.
